core_lib/src/util/util.cpp

```diff
@@ -126,3 +128,59 @@ QString uniqueString(int len)
     s[len] = 0;
     return QString::fromUtf8(s);
 }
+
+QString validateDataPath(QString filePath, QString dataDirPath)
+{
+    // Make sure src path is relative
+    if (!QFileInfo(filePath).isRelative()) return QString();
+
+    QFileInfo fi(dataDirPath, filePath);
+    // Recursively resolve symlinks
+    QString canonicalPath = fi.canonicalFilePath();
+
+    // Iterate over parent directories of the file path to see if one of them equals the data directory
+    QDir dataDir(dataDirPath);
+    if (dataDir.exists()) dataDir.setPath(dataDir.canonicalPath());
+    if (canonicalPath.isEmpty())
+    {
+        // File does not exist, use absolute path and attempt to resolve symlinks again for each parent directory
+        fi.setFile(fi.absoluteFilePath());
+        QDir ancestor(fi.absoluteFilePath());
+        while (true) {
```

Comment:
I don't quite understand this while loop. Could you please elaborate a bit more? probably with some examples?

Reply:
For an example, let's say the data directory is at the path `/a/b/c` and the file path being validated is at `/a/b_symlink/c/d/file.txt`. The directory `b_symlink` is obviously a symlink to `/a/b`. We will also assume the directory `/a/b/c/d` and `file.txt` do not exist and directories `/a`, `/a/b`, and `/a/b/c` do exist. I will be walking through the logic based on the slightly rearranged code I have pushed since your review so pull the latest code before following along. 

This loop iterates up the directory hierarchy of the file path. Keep in mind that it only runs if the canonical path of the file is empty, which means the file and possibly some of its parent directories don't exist or are dangling symlinks (symlinks where the target location does not exist, QDir::exists() also returns false in this case).

Iteration 1:
1. ancestor is `/a/b_symlink/c/d/file.txt`
2. ancestor is not equal to data dir
3. newAncestor is set to the parent of ancestor: `/a/b_symlink/c/d`
4. newAncestor does not exist
5. newAncestor is different from ancestor
6. Set ancestor to newAncestor

Iteration 2:
1. ancestor is `/a/b_symlink/c/d`
2. ancestor is not equal to data dir
3. newAncestor is set to the parent of ancestor: `/a/b_symlink/c`
4. newAncestor does exist, so set new ancestor to the canonical path, which resolves the symlink b_symlink. newAncestor is now: `/a/b/c`
5. newAncestor is different from ancestor
6. Set ancestor to newAncestor

Iteration 3:
1. ancestor is `/a/b/c`
2. ancestor is equal to data dir, that means the file would be within the data directory after resolving all symlinks  if it and all of its parent directories existed. As a positive result, return the file path.

If the data dir was not `/a/b` and instead something like `/x`, then the loop would continue with ancestor = `/a/b` and then `/a` and then `/`. When ancestor is the root directory, calling dir on it will again return the root directory `/` and newAncestor will equal ancestor. As a negative result, return an invalid string because we now know that the file would not be within the data directory after resolving symlinks if it and all of its parent directories existed.

Reply:
Alright, thanks for the thorough explanation.

I now understand the nuance here. The condition `if (ancestor == newAncestor)` indicates that the path has reached the root directory, as the parent directory of the root is itself. This wasn’t immediately obvious to me, so I couldn't grasp it at first.

Would you be able to put this explanation into the code as a comment or a util function so the code can express its intent clearly? can `QDir::isRoot()` be used here?